File: LaboReseauxBinome/Evaluation5/Application_CIAChat/Message.cpp
#include "Message.h"

Message::Message(int messageType, const std::string &user, const std::string &message, const std::string &tag)
		: messageType(messageType), message(message), tag(tag), user(user) {
	this->digest = Message::createDigest(this->message);
	this->time = Message::getCurrentTime();
}

Message::Message(int messageType, const std::string &user, const std::string &message, const std::string &tag, int digest)
		: messageType(messageType), message(message), tag(tag), digest(digest), user(user) {
	this->time = Message::getCurrentTime();
}

Message::Message(int messageType, const std::string &user, const std::string &time, const std::string &message, const std::string &tag, int digest)
				: time(time), messageType(messageType), message(message), tag(tag), digest(digest), user(user) {}

int Message::getMessageType() const { return this->messageType; }
const std::string &Message::getTime() const { return time; }
const std::string &Message::getMessage() const { return message; }
const std::string &Message::getTag() const { return tag; }
int Message::getDigest() const { return digest; }
const std::string &Message::getUser() const { return user; }

std::ostream &operator<<(std::ostream &os, const Message &message) {
	os << "[" << message.time << "]";
	os << "[" << message.getMessageTypeAsString() << "]";
	os << "[" << message.tag << "]";
	os << " " << message.user << " :";
	os << " " << message.message;

	return os;
}

std::string Message::getMessageTypeAsString() const {
	switch(this->messageType){
		case POST_QUESTION:
			return std::string("Question");
		case ANSWER_QUESTION:
			return std::string("Réponse");
		case POST_EVENT:
			return std::string("Event");
		default:
			return std::string();
	}
}

int Message::createDigest(std::string msg){
	int hash = 0;

	for(int i=0; i<msg.size(); i++)
		hash += msg.at(i);

	return hash%67;
}

std::string Message::getCurrentTime() {
	std::stringstream buffer;
	time_t tt = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
	struct tm *ptm = localtime(&tt);
	buffer << std::put_time(ptm, "%H:%M:%S");

	return buffer.str();
}

std::string Message::generateTag(std::vector<Message> messageList) {
	bool dispo = false;
	int random;
	srand(static_cast<unsigned int>(std::time(nullptr)));

	while(!dispo) {
		dispo = true;
		random = rand() % 10000 - 1;

		for (int i = 0; i<messageList.size() && dispo; ++i) {
			dispo = (messageList[i].tag) == std::to_string(random);
		}
	}

	return std::to_string(random);
}

File: LaboReseauxBinome/Evaluation5/IACOP/src/network/Message.java
package network;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;

public class Message {
	public static enum UDP {
		POST_QUESTION(0),
		ANSWER_QUESTION(1),
		POST_EVENT(2);

		private int value;

		private UDP(int value){
			this.value = value;
		}

		private int getValue(){
			return this.value;
		}
	};

	private final String time;
	private final UDP messageType;
	private final String message;
	private final String tag;
	private final int digest;
	private final String user;

	public Message(UDP messageType, String user, String time, String tag, String message, int digest) {
		this.time = time;
		this.messageType = messageType;
		this.message = message;
		this.tag = tag;
		this.digest = digest;
		this.user = user;
	}

	public Message(UDP messageType, String user, String tag, String message, int digest) {
		DateFormat dateFormat  = new SimpleDateFormat("HH:mm:ss");
		Calendar calendar = Calendar.getInstance();

		this.time = dateFormat.format(calendar.getTime());
		this.messageType = messageType;
		this.message = message;
		this.tag = tag;
		this.digest = digest;
		this.user = user;
	}

	public static boolean isMessageType(int type){
		for(UDP udp: UDP.values()){
			if(udp.getValue() == type)
				return true;
		}
		return false;
	}

	@Override
	public String toString() {
		// message = [12:47:31][messateType][tag] laurent : voici mon message!
		StringBuilder stringBuilder = new StringBuilder();

		stringBuilder.append("[");
		stringBuilder.append(this.time);
		stringBuilder.append("][");

		switch(this.messageType){
			case POST_QUESTION:
				stringBuilder.append("Question");
				break;
			case ANSWER_QUESTION:
				stringBuilder.append("Réponse");
				break;
			case POST_EVENT:
				stringBuilder.append("Event");
				break;
		}
		stringBuilder.append("][");
		stringBuilder.append(this.tag);
		stringBuilder.append("] ");
		stringBuilder.append(this.user);
		stringBuilder.append(" : ");
		stringBuilder.append(this.message);

		return stringBuilder.toString();
	}

	//<editor-fold defaultstate="collapsed" desc="GETTER & SETTER">
	public String getTime() {
		return time;
	}
	public UDP getMessageType() {
		return messageType;
	}
	public int getMessageTypeInt(){
		return messageType.getValue();
	}
	public String getMessage() {
		return message;
	}
	public String getTag() {
		return tag;
	}
	public int getDigest() {
		return digest;
	}
	public String getUser() {
		return user;
	}
	//</editor-fold>
}
